Add structural OpenAPI schema to Tekton CRDs

burigolucas opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Address [Issue 1461](https://github.com/tektoncd/pipeline/issues/1461) by implementing a script to update the OpenAPI schema to Tekton CRDs using controller-gen and adding the updated CRDs with OpenAPI schema.

Revise unit tests and example YAML files to address issues caused by the now enabled API server schema validation.

Note that the current script `hack/update-schemas.sh` return several errors/warnings and status code 1 from execution of the controller-gen CLI due to issues when parsing the schema from the source code. Besides, processing multiple files at once is causing segmentation fault. Therefore, in the script, each file is processed separately. In addition, FIXME notes are added to be addressed in the future once new features become available in controller-gen.

Closes #1461 

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
action required: The structural OpenAPI schema to Tekton CRDs are added enabling API server schema validation and supporting `kubectl explain` to describe fields and structure of Tekton CRDs. Due to the API server schema validation, users should make sure Tekton CRs have a valid schema when creating or updating CRs.
```

burigolucas commented:
/kind documentation

vdemeester commented:
/ok-to-test

burigolucas commented:
Hi @vdemeester, I have identified issues with the CRDs. A few of them were invalid due to recursive fields. Following the fix proposed in this issue [Generating CRD manifest for struct with recursive field](https://github.com/kubernetes-sigs/controller-tools/issues/585), I have added the markers 
```
	// +kubebuilder:pruning:PreserveUnknownFields
	// +kubebuilder:validation:Schemaless
```
in the affected fields and regenerated the CRDs. I have also disabled the embedded object meta as it does not provide useful schema in my opinion and only makes the files more bloated.

AFAIK, there is still one issue to be solved, namely, the CRD for the `PipelineRun` is too large. We reach the file limit with `kubectl create`. Any ideas how best to address that?

vdemeester commented:
```
Error from server (Invalid): error when creating "STDIN": CustomResourceDefinition.apiextensions.k8s.io "pipelines.tekton.dev" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
Error from server (RequestEntityTooLarge): error when creating "STDIN": Request entity too large: limit is 3145728
Error from server (Invalid): error when creating "STDIN": CustomResourceDefinition.apiextensions.k8s.io "tasks.tekton.dev" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
Error from server (Invalid): error when creating "STDIN": CustomResourceDefinition.apiextensions.k8s.io "taskruns.tekton.dev" is invalid: metadata.annotations: Too long: must have at most 262144 bytes
```

vdemeester commented:
@burigolucas could we refer to another schema ? `PipelineRun` gets too big because of `pipelineSpec` that is, well the `Pipeline.Spec` field (and that one also contains `Task.Spec` with `taskSpec`). Maybe we should "skip" those "embedded" type in the schema and instead "document" or "refer" to the other schema ?

vdemeester commented:
> Are you trying `kubectl apply`? In that case it will fail as it cannot store the changes in the annotations. It should work with `kubectl create` except for the `PipelineRun` due to the size. 

Yeah it's what (`kubectl apply`) the CI (github workflow) does (and most our users as well 😅 )

> If you remove the older version of the `PipelineRun` in the CRD, then you can also verify that the CRD is valid. Perhaps we could omit parts of the schema of the `PipelineRun` for the older CRD version?

You mean the `v1beta1` part right ?

vdemeester commented:
I would be fine only having the shema for the `v1` (or the highest) version available in the CRD to be honest.
@afrittoli @tektoncd/core-maintainers what do you think ?

burigolucas commented:
I have just tried to remove the pipelineSpec from the schema of the pipelineRun and it is now small enough, even with both versions of the CRD. But we might need to reduce other CRDs further if we need to support `kubectl apply` due to the limits of the annotation field.

burigolucas commented:
Remove also the schema of TaskSpec in TaskRun and added a comment for the user to refer to the schema of Task. This also reduced size but currently `Task`, `TaskRun`, `Pipeline` and `PipelineRun` are all too large to be stored in the annotations field when using `kubectl apply`. Any further suggestions what else to remove from the schema to address the length of the files?

vdemeester commented:
> Removed also the schema of `TaskSpec` in `TaskRun` and added a comment for the user to refer to the schema of Task. This also reduced size but currently `Task`, `TaskRun`, `Pipeline` and `PipelineRun` are all too large to be stored in the annotations field when using `kubectl apply`. Any further suggestions what else to remove from the schema to address the length of the files?

I am "fine" with requiring things to be done using `kubectl create` and document this I guess ?
cc @afrittoli

afrittoli commented:
> > Removed also the schema of `TaskSpec` in `TaskRun` and added a comment for the user to refer to the schema of Task. This also reduced size but currently `Task`, `TaskRun`, `Pipeline` and `PipelineRun` are all too large to be stored in the annotations field when using `kubectl apply`. Any further suggestions what else to remove from the schema to address the length of the files?
> 
> I am "fine" with requiring things to be done using `kubectl create` and document this I guess ? cc @afrittoli

This would be for the CRDs only, right? 
Existing installation scripts and tools may break if we make this change. To provide for a smoother transition, we could add an extra release file to our releases:
- `release.yaml`: includes sha and tag for images, includes CRDs with schema
- `release.notag.yaml`: includes CRDs with schema
- `release.oldcrds.yaml`: includes sha and tag for images, includes CRDs without schema

This way users could point to `release.oldcrds.yaml` until they migrate to installing via `kubectl create`.

vdemeester commented:
@afrittoli should it be the other way around for a while. Like next release we have a `release-with-schemas.yaml` and we announce a given timeline for changing the default and then we do the other way around ?

afrittoli commented:
> @afrittoli should it be the other way around for a while. Like next release we have a `release-with-schemas.yaml` and we announce a given timeline for changing the default and then we do the other way around ?

Sure, we can do that, I was trying to avoid too many changes :D but it's safer the way you proposed

burigolucas commented:
Should we then keep the current CRDs untouched and store the new generated CRDs with schemas in a different path?

afrittoli commented:
> Should we then keep the current CRDs untouched and store the new generated CRDs with schemas in a different path?

If possible, I would prefer to have one version of the CRDs in the repos and, at release time:
- copy `release.yaml` to `release.schemas-in-crds.yaml`
- strip the schemas from `release.yaml`

@vdemeester @burigolucas WDYT?

vdemeester commented:
> > Should we then keep the current CRDs untouched and store the new generated CRDs with schemas in a different path?
> 
> If possible, I would prefer to have one version of the CRDs in the repos and, at release time:
> 
>     * copy `release.yaml` to `release.schemas-in-crds.yaml`
> 
>     * strip the schemas from `release.yaml`
> 
> 
> @vdemeester @burigolucas WDYT?

I agree, but it means we need to change our CI script to use `create` instead of `apply` 👼🏼 .

burigolucas commented:
@vdemeester, @afrittoli  I have analyzed the schema more carefully and identified good candidates to omit from the schema (e.g., Pod.spec.volumes, Pod.spec.affinity, PersistentVolumeClaim). While omiting these schema, I have added a note in the description for the user where to located the schema in the k8s API. With these changes, the CRDs are now significantly smaller and can be `kubectl apply`ed. In my opinion, this is a reasonable solution and do not require different release files. However the e2e tests are broken. I will need help to address that.

afrittoli commented:
> @vdemeester, @afrittoli I have analyzed the schema more carefully and identified good candidates to omit from the schema (e.g., Pod.spec.volumes, Pod.spec.affinity, PersistentVolumeClaim). While omitting these schema, I have added a note in the description for the user where to locate the schema in the K8s API. With these changes, the CRDs are now significantly smaller and can be `kubectl apply`ed. In my opinion, this is a reasonable solution and does not require different release files. However, the e2e tests are broken. I will need help to address that.

Thank you for this work @burigolucas - this sounds like a great solution to me

burigolucas commented:
One of the issues with the e2e is related to schema for fields which can have different types. E.g., the schema for `Pipeline.spec.params.default` is set by controller-gen as `object`, throwing a validation error when attempting to create a pipeline with a string value as the default param value. I will revise the CRDs schema again to try resolving this and other related issues.

burigolucas commented:
I have added the schema for the feature flags as well and had to update unit tests and a few YAML examples due to the API server schema validation. This validation will cause breaking changes to users who were not creating CRs with a valid schema. Besides, the schema for feature flags has changed with the fields starting with lower case instead of upper case. I am not able to run the e2e tests locally to fully test though.

burigolucas commented:
/test pull-tekton-pipeline-go-coverage-df

burigolucas commented:
@vdemeester, @afrittoli I have revised all the previously falling e2e tests and made the required changes to fix the issues. I have also rebased the branch to a address a merge conflict and squashed all my commits to a single one as suggested in the documentation for contributors. I do not know to to fix the failing job pull-tekton-pipeline-go-coverage-df though. Could one of you please approve the workflows to check if all tests are now passing? Thanks in advance!

afrittoli commented:
/test pull-tekton-pipeline-go-coverage-df

afrittoli commented on hack/update-schemas.sh:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 The Tekton Authors
```
NIT: this should be 2025

vdemeester commented:
@burigolucas ah there is one lint issue it seems.

afrittoli commented on examples/v1/taskruns/beta/stepaction-passing-results.yaml:
```diff
@@ -60,7 +60,7 @@ kind: TaskRun
   name: step-action-run
 spec:
-  TaskSpec:
+  taskSpec:
```
Is this example not executed in CI, or is it executed but the capital letter version was accepted somehow by Tekton?

afrittoli commented on hack/update-codegen.sh:
```diff
@@ -109,3 +109,6 @@ ${REPO_ROOT_DIR}/hack/update-openapigen.sh
 ${REPO_ROOT_DIR}/hack/update-reference-docs.sh
+
+# Add structural OpenAPI schema to Tekton CRDs
+${REPO_ROOT_DIR}/hack/update-schemas.sh
```
After running this script, we need to verify that the generated CRDs have not been modified. 
The `verify-codegen.sh` script achieves this for the other code gens, but it does not include the `config` folder, so it needs to be extended, otherwise we risk missing schema updates.

afrittoli commented:
> @burigolucas ah there is one lint issue it seems.

It's not in a file changed by this PR though, which should not happen... @vdemeester any idea why this happened?

vdemeester commented:
> > @burigolucas ah there is one lint issue it seems.
> 
> It's not in a file changed by this PR though, which should not happen... @vdemeester any idea why this happened?

I need to check, but I think the github action checks all files by default (not just the changed one).

vdemeester commented:
@afrittoli See https://github.com/tektoncd/pipeline/pull/8604 🙃

burigolucas commented:
> > > @burigolucas ah there is one lint issue it seems.
> > 
> > 
> > It's not in a file changed by this PR though, which should not happen... @vdemeester any idea why this happened?
> 
> I need to check, but I think the github action checks all files by default (not just the changed one).

I remvoed the unsed directive in the latest commit. Should I revert it back then?

vdemeester commented:
> This looks amazing!
> 
> Now, I am sorry for the scope-creep but would it be possible to include the `x-kubernetes-patch-strategy` and `x-kubernetes-patch-merge-key` [extensions](https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/README.md#x-kubernetes-patch-strategy-and-x-kubernetes-patch-merge-key) for the same objects that we have `x-kubernetes-list-type`. This would help enable strategic merge patch using the [openapi feature](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/openapi/) of kustomize.
> 
> (It is just a suggestion and I understand that you might prefer to keep it for a separate PR)

Given the back and forth on the PR, I would suggest we do this in a follow-up PR, but it's up to you @burigolucas.

> @vdemeester I removed the unused directive in the latest commit. Should I rather revert it back then?

If the CI is green, I think it's fine to keep as is 😉

burigolucas commented:
@vdemeester it looks like we are having rate limiting issues with DockerHub (See the extracts of the logs below for each of the 4 failed tests). Is this something new? Could we simply retry the failed set of tests to check for flaky results?

- `TestExamples/v1/taskruns/dind-sidecar`
```verbatim
message: 'the sidecar "server" in TaskRun "dind-sidecar-882ss" failed to pull
              the image "". The pod errored with the message: "Back-off pulling image "docker@sha256:74e78208fc18da48ddf8b569abe21563730845c312130bd0f0b059746a7e10f5"."'
```

- `TestExamples/v1/pipelineruns/pipelinerun`
```verbatim
2025-03-06T16:45:05.2945141Z         >>> Pod demo-pipeline-run-1-build-image-pod Container step-build-and-push:
2025-03-06T16:45:05.2945967Z         [36mINFO[0m[0000] Resolved base name golang:1.18 to builder    
2025-03-06T16:45:05.2946570Z         [36mINFO[0m[0000] Retrieving image manifest golang:1.18        
2025-03-06T16:45:05.2947300Z         [36mINFO[0m[0000] Retrieving image golang:1.18 from registry index.docker.io 
2025-03-06T16:45:05.2949243Z         error building image: GET https://index.docker.io/v2/library/golang/manifests/1.18: TOOMANYREQUESTS: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit
```

- `TestExamples/v1/taskruns/step-by-digest`
```verbatim
-03-06T16:45:14.6943112Z           - lastTransitionTime: "2025-03-06T16:45:14Z"
2025-03-06T16:45:14.6944275Z             message: 'failed to create task run pod "step-by-digest-2c4dw": translating TaskSpec
2025-03-06T16:45:14.6947703Z               to Pod: GET https://index.docker.io/v2/library/busybox/manifests/sha256:1303dbf110c57f3edf68d9f5a16c082ec06c4cf7604831669faf2c712260b5a0:
2025-03-06T16:45:14.6950438Z               TOOMANYREQUESTS: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit.
2025-03-06T16:45:14.6952083Z               Maybe invalid TaskSpec'
2025-03-06T16:45:14.6952909Z             reason: PodCreationFailed
2025-03-06T16:45:14.6953425Z             status: "False"
2025-03-06T16:45:14.6953886Z             type: Succeeded
```

- `TestExamples/v1/taskruns/entrypoint-resolution`
```verbatim
2025-03-06T16:46:58.1804050Z           - lastTransitionTime: "2025-03-06T16:46:57Z"
2025-03-06T16:46:58.1805171Z             message: 'failed to create task run pod "entrypoint-resolution": translating TaskSpec
2025-03-06T16:46:58.1806162Z               to Pod: GET https://index.docker.io/v2/amd64/ubuntu/manifests/latest: TOOMANYREQUESTS:
2025-03-06T16:46:58.1807221Z               You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit.
2025-03-06T16:46:58.1808371Z               Maybe invalid TaskSpec'
2025-03-06T16:46:58.1808760Z             reason: PodCreationFailed
2025-03-06T16:46:58.1809364Z             status: "False"
2025-03-06T16:46:58.1809680Z             type: Succeeded
```

burigolucas commented:
/retest

112358fn commented:
@burigolucas does [this issue description](https://github.com/tektoncd/pipeline/issues/8628) looks ok?

vdemeester commented:
/retest

abayer commented:
/lgtm
